src/ring_theory/jacobson.lean

```diff
@@ -252,4 +259,167 @@ end
+        refl } },
+    { obtain ⟨p, rfl⟩ := quotient.mk_surjective p',
+      refine polynomial.induction_on p
+        (λ r, subring.subset_closure $ set.mem_image_of_mem _ (or.inr degree_C_le))
+        (λ _ _ h1 h2, _) (λ n _ hr, _),
+      { convert subring.add_mem _ h1 h2,
+        rw [ring_hom.map_add, ring_hom.map_add] },
+      { rw [pow_succ X n, mul_comm X, ← mul_assoc, ring_hom.map_mul, ϕ'.to_map.map_mul],
+        exact subring.mul_mem _ hr (subring.subset_closure (set.mem_image_of_mem _ (or.inl rfl))) } } }
+end
+
+theorem is_jacobson_polynomial_iff_is_jacobson {R : Type*} [comm_ring R] : is_jacobson (polynomial R) ↔ is_jacobson R :=
+begin
+  split; introI H,
+  { exact is_jacobson_of_surjective ⟨eval₂_ring_hom (ring_hom.id _) 1, λ x, ⟨C x, by simp⟩⟩ },
+  { rw is_jacobson_iff_prime_eq,
+    intros I hI,
+    let R' := ((quotient.mk I).comp C).range,
+    let i : R →+* R' := ((quotient.mk I).comp C).range_restrict,
+    have hi : function.surjective (i : R → R') := ((quotient.mk I).comp C).surjective_onto_range,
+    have hi' : (polynomial.map_ring_hom i : polynomial R →+* polynomial R').ker ≤ I,
+    { refine λ f hf, polynomial_mem_ideal_of_coeff_mem_ideal I f (λ n, _),
+      rw [mem_comap, ← quotient.eq_zero_iff_mem, ← ring_hom.comp_apply],
+      rw [ring_hom.mem_ker, coe_map_ring_hom] at hf,
+      replace hf := congr_arg (λ (f : polynomial R'), f.coeff n) hf,
+      simp only [coeff_map, coeff_zero] at hf,
+      rwa [subtype.ext_iff, ring_hom.coe_range_restrict] at hf },
+    haveI hR' : is_jacobson R' := is_jacobson_of_surjective ⟨i, hi⟩,
+    let I' : ideal (polynomial R') := I.map (polynomial.map_ring_hom i),
+    haveI : I'.is_prime := map_is_prime_of_surjective (polynomial.map_surjective i hi) hi',
+    suffices : (I.map (polynomial.map_ring_hom i)).jacobson = (I.map (polynomial.map_ring_hom i)),
+    { replace this := congr_arg (comap (polynomial.map_ring_hom i)) this,
+      rw [← map_jacobson_of_surjective _ hi',
+        comap_map_of_surjective _ _, comap_map_of_surjective _ _] at this,
+      refine le_antisymm (le_trans (le_sup_left_of_le le_rfl)
+        (le_trans (le_of_eq this) (sup_le le_rfl hi'))) le_jacobson,
+      all_goals {exact polynomial.map_surjective i hi} },
+    refine is_jacobson_polynomial_of_domain R' I' (eq_zero_of_polynomial_mem_map_range I hi') },
+end
+
+instance [is_jacobson R] : is_jacobson (polynomial R) :=
+is_jacobson_polynomial_iff_is_jacobson.mpr ‹is_jacobson R›
+
+/-- General form of the nullstellensatz for jacobson rings, since in a jacobson ring we have
+  `Inf {P maximal | P ≥ I} = Inf {P prime | P ≥ I} = I.radical`. Fields are always jacobson,
+  and in that special case this is (most of) the classical nullstellensatz,
+  since `I(V(I))` is the intersection of maximal ideals containing `I`, which is then `I.radical` -/
+lemma is_jacobson_mv_polynomial_fin [H : is_jacobson R] (n : ℕ) :
+  is_jacobson (mv_polynomial (fin n) R) :=
+nat.rec_on n
```

Comment:
Is there a reason why you don't use the equation complier here?
I believe it would be slightly more readable.

Reply:
The only reason is that I had forgotten lean had nice pattern matching syntax like that. I've switched it now